core/unit_test/TestViewTypedefs.cpp

```diff
@@ -0,0 +1,210 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+
+namespace {
+
+// clang-format off
+template<class T>
+struct data_analysis {
+  using data_type = T;
+  using const_data_type = const T;
+  using runtime_data_type = T;
+  using runtime_const_data_type = const T;
+  using non_const_data_type = std::remove_const_t<T>;
+};
+
+template<class T>
+struct data_analysis<T*> {
+  using data_type = typename data_analysis<T>::data_type*;
+  using const_data_type = typename data_analysis<T>::const_data_type*;
+  using runtime_data_type = typename data_analysis<T>::runtime_data_type*;
+  using runtime_const_data_type = typename data_analysis<T>::runtime_const_data_type*;
+  using non_const_data_type = typename data_analysis<T>::non_const_data_type*;
+};
+
+template<class T, size_t N>
+struct data_analysis<T[N]> {
+  using data_type = typename data_analysis<T>::data_type[N];
+  using const_data_type = typename data_analysis<T>::const_data_type[N];
+  using runtime_data_type = typename data_analysis<T>::runtime_data_type*;
+  using runtime_const_data_type = typename data_analysis<T>::runtime_const_data_type*;
+  using non_const_data_type = typename data_analysis<T>::non_const_data_type[N];
+};
+
+
+template<class V, class VT, class T, class L, class S, class M, class HostMirrorSpace, class ValueType, class ReferenceType>
+constexpr bool test_view_typedefs_impl() {
+  using v_t = V;
+  // ========================
+  // inherited from ViewTraits
+  // ========================
+  static_assert(std::is_same_v<typename v_t::data_type, T>);
+  static_assert(std::is_same_v<typename v_t::const_data_type, typename  data_analysis<T>::const_data_type>);
+  static_assert(std::is_same_v<typename v_t::non_const_data_type, typename  data_analysis<T>::non_const_data_type>);
+  
+  // these should be deprecated and for proper testing we need embedded stupid type
```

Comment:
No idea what this means

Reply:
these things only differ from the previous typedefs for stupid embedded types

Reply:
what's an embedded type?

Reply:
reworked the comment: its the ensemble types with a hidden view dimension